src/components/app/routes/data/utils.js

```diff
@@ -60,46 +60,48 @@ export async function ensureEnterpriseAppData({
-              (license) => license.uuid === activatedOrAutoAppliedLicense.uuid,
-            );
-            const isCurrentStatusMatchingLicenseStatus = status === activatedOrAutoAppliedLicense.status;
-            if (licensesIncludesActivatedOrAutoAppliedLicense) {
-              updatedLicensesByStatus[status] = isCurrentStatusMatchingLicenseStatus
-                ? licenses.filter((license) => license.uuid !== activatedOrAutoAppliedLicense.uuid)
-                : [...licenses, activatedOrAutoAppliedLicense];
-            } else if (isCurrentStatusMatchingLicenseStatus) {
-              updatedLicensesByStatus[activatedOrAutoAppliedLicense.status].push(activatedOrAutoAppliedLicense);
-            }
-          });
-          // Optimistically update the query cache with the auto-activated or auto-applied subscription license.
-          const updatedSubscriptionLicenses = subscriptionsData.subscriptionLicenses.length > 0
-            ? subscriptionsData.subscriptionLicenses.map((license) => {
-            // Ensures an auto-activated license is updated in the query cache to change
-            // its status from "assigned" to "activated".
-              if (license.uuid === activatedOrAutoAppliedLicense.uuid) {
-                return activatedOrAutoAppliedLicense;
-              }
-              return license;
-            })
-            : [activatedOrAutoAppliedLicense];
+          const { subscriptionLicensesByStatus, subscriptionLicenses } = subscriptionsData;
+          // Create a deep copy of the structure using .map for immutability, removing
+          // the `activatedOrAutoAppliedLicense` from each list. Then, re-add the license
+          // to the correct status list.
+          const updatedLicensesByStatus = Object.fromEntries(
+            Object.entries(subscriptionLicensesByStatus).map(([key, licenses]) => [
+              key,
+              licenses.filter(
+                (existingLicense) => existingLicense.uuid !== activatedOrAutoAppliedLicense.uuid,
+              ), // Remove license immutably
+            ]),
+          );
+          if (!updatedLicensesByStatus[activatedOrAutoAppliedLicense.status]) {
+            updatedLicensesByStatus[activatedOrAutoAppliedLicense.status] = [];
+          }
+          updatedLicensesByStatus[activatedOrAutoAppliedLicense.status].push(activatedOrAutoAppliedLicense);
 
+          // Update the flat subscription licenses list
+          const updatedSubscriptionLicenses = subscriptionLicenses.some(
+            (existingLicense) => existingLicense.uuid === activatedOrAutoAppliedLicense.uuid,
+          )
+            ? subscriptionLicenses.map((existingLicense) => (
+              existingLicense.uuid === activatedOrAutoAppliedLicense.uuid
+                ? activatedOrAutoAppliedLicense
+                : existingLicense
+            ))
+            : [...subscriptionLicenses, activatedOrAutoAppliedLicense];
```

Comment:
[nit/non-blocking] Logic here is a little dense, might be worth breaking it apart a bit.

Reply:
Discussed in Slack. Will update as follows:

```jsx
// Update the flat subscription licenses list
const updatedSubscriptionLicenses = [...subscriptionLicenses];
const licenseIndex = subscriptionLicenses.findIndex(
  (license) => license.uuid === activatedOrAutoAppliedLicense.uuid,
);
if (licenseIndex >= 0) {
  // Replace the existing license
  updatedSubscriptionLicenses[licenseIndex] = activatedOrAutoAppliedLicense;
} else {
  // Add the new license
  updatedSubscriptionLicenses.push(activatedOrAutoAppliedLicense);
}
```